src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -175,101 +217,28 @@ private static bool CaretIsInForStatementCondition(int caretPosition, ForStateme
-        {
-            // work our way up the tree, looking for a node of interest within the current statement
-            var nodeFound = false;
-            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
-            {
-                // This feature operates inside the following types of nodes, 
-                // and can be expanded if more cases are implemented:
-                // ArgumentList: covers method invocations like object.Method(arg)
-                // ArrayRankSpecifier: covers new Type[dim]
-                // ElementAccessExpression: covers indexer invocations like array[index]
-                // ParenthesizedExpression: covers (3*(x+y))
-                if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.BracketedArgumentList, SyntaxKind.ParenthesizedExpression))
-                {
-                    nodeFound = true;
-                }
-
-                // No special action is performed at this time if `;` is typed inside a string, including
-                // interpolated strings.  
-                if (IsInAString(currentNode, caret))
-                {
-                    return false;
-                }
-
-                // We reached the root without finding a statement
-                if (currentNode.Parent == null)
-                {
-                    return false;
-                }
+                caretPosition <= forStatementSyntax.Initializers.Span.End;
 
-                currentNode = currentNode.Parent;
-            }
+        private static bool CaretIsInForStatementIterator(int caretPosition, ForStatementSyntax forStatementSyntax)
+           => caretPosition > forStatementSyntax.Incrementors.Span.Start &&
+               caretPosition <= forStatementSyntax.Incrementors.Span.End;
 
-            // if we never found a statement, or a node of interest, or the statement kind is not a candidate for completion, return
-            if (currentNode == null || !nodeFound || !StatementIsACandidate(currentNode, caret.Position))
-            {
-                return false;
-            }
-
-            return true;
-        }
-
-        private static bool IsStatementOrFieldDeclaration(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
-            => syntaxFacts.IsStatement(currentNode) || currentNode.IsKind(SyntaxKind.FieldDeclaration);
-
-        private static bool IsInAString(SyntaxNode currentNode, SnapshotPoint caret)
+        private static bool IsInAStringOrCharacter(SyntaxNode currentNode, SnapshotPoint caret)
             // Check to see if caret is before or after string
-            => currentNode.IsKind(SyntaxKind.InterpolatedStringExpression, SyntaxKind.StringLiteralExpression)
+            => currentNode.IsKind(SyntaxKind.InterpolatedStringExpression, SyntaxKind.StringLiteralExpression, SyntaxKind.CharacterLiteralExpression)
                 && caret.Position < currentNode.Span.End
```

Comment:
`<=` here as well?

Reply:
This one doesn't need to be <= because we don't want to return true if the caret is outside of the string or character.

The other methods that I changed to <= are now called after the caret might have been moved to the end, so need to include the endpoint.